### PR TITLE
Notifications for remote shares

### LINKED ISSUE #19587: Replace existing remote share user interaction to interactive notifications
As an admin, I want to be able to have users accept remote share links with an interactive notification so that only desired content is shared among users across instances.

(We are moving accept remote shares to an interactive dialog, and to replace the current implementation)

Acceptance Criteria:
- Migrate the code to the new interactive notifications
- Note: If notification app is off, remote shares cannot be accepted.

### PR DESCRIPTION
**Pull request by nickvergessen**
Functionally only visible with https://github.com/owncloud/notifications/pull/62

Fix #19587

@rullzer @PVince81 please review

### CODE REVIEW

File: apps/files_sharing/js/external.js [hunk 1]
```diff
@@ -69,9 +69,21 @@
...
 			this.filesApp = filesApp;
 			this.processIncomingShareFromUrl();
-			this.processSharesToConfirm();
+
+			if (!$('#header').find('div.notifications').length) {
+				// No notification app, display the modal
+				this.processSharesToConfirm();
+			}
+
+			$('body').on('OCA.Notification.Action', function(e) {
```
Comment (PVince81) [inline apps/files_sharing/js/external.js]: you could put that in an else block (optional)

File: apps/files_sharing/js/external.js [hunk 2]
```diff
@@ -69,9 +69,21 @@
...
+				if (e.notification.app === 'files_sharing' && e.notification.object_type === 'remote_share' && e.action.type === 'POST') {
```
Comment (PVince81) [inline apps/files_sharing/js/external.js]: camel case (object_type) is usually not accepted in the JS coding style, but I guess this comes from the original JSON structure, so ok to keep

**Comment by PVince81** [comment]
Code looks good, as discussed, saw it working on your screen :+1:

File: apps/files_sharing/js/external.js (hunk 2 above)
Comment (nickvergessen) [inline apps/files_sharing/js/external.js]: yes, that's json

**Comment by rullzer** [comment]
Works!
Code looks good. :+1:

**Comment by PVince81** [comment]
Related Oracle fail:
```
11:55:26 1) Test_Files_Sharing_S2S_OCS_API::testCreateShare
11:55:26 Failed asserting that false is true.
11:55:26 
11:55:26 /ssd/jenkins/workspace/core-ci-linux/database/oci/label/SLAVE/apps/files_sharing/tests/server2server.php:112
```

**Comment by PVince81** [comment]
Could it be the lastInsertId https://github.com/owncloud/core/pull/22253/files#diff-095b5f4e7ee9665a242886c86670b4b8R84 ?

Oracle is known to dislike those...

File: apps/files_sharing/api/server2server.php [hunk 3]
```diff
@@ -81,6 +81,7 @@ public function createShare($params) {
 
 			try {
 				$externalManager->addShare($remote, $token, '', $name, $owner, false, $shareWith, $remoteId);
+				$shareId = \OC::$server->getDatabaseConnection()->lastInsertId('`*PREFIX*share_external`');
```
Comment (PVince81) [inline apps/files_sharing/api/server2server.php]: might be possible to get the share item from the shared storage, I see that `addShare` returns a `IMountPoint` instance:
`$mount->getStorage()->getShare()['id']` (I didn't test, just guessed from the code)

Comment (nickvergessen) [inline apps/files_sharing/api/server2server.php]: Its null here

Comment (rullzer) [inline apps/files_sharing/api/server2server.php]: External shares don't have that. I believe we just mount them as 'true' webdav mounts. @schiesbn

Comment (rullzer) [inline apps/files_sharing/api/server2server.php]: Couldn't we just add getShareId to the ISharedStorge interface? The data is there when we construct the storage..

Comment (rullzer) [inline apps/files_sharing/api/server2server.php]: aaah no because we first insert it temp... mmm stupid oci...

Comment (PVince81) [inline apps/files_sharing/api/server2server.php]: @rullzer ISharedStorage already has getShare() which returns the old-school share array that has the id in it

Comment (rullzer) [inline apps/files_sharing/api/server2server.php]: https://github.com/owncloud/core/blob/master/apps/files_sharing/lib/isharedstorage.php <-- well that is empty ;)

But the thing is that we don't get a mountpoint on the original insert. So there is no way to get an id from it...

Comment (PVince81) [inline apps/files_sharing/api/server2server.php]: Why would it be empty ? Does it mean it's a special case where no notification should be created ?
Or is it only that unit test that is weird ? (it creates a share with `null`, is that even a thing ?)

Comment (nickvergessen) [inline apps/files_sharing/api/server2server.php]: Woops, the table name should not be quoted in lastInsertId, I guess that will solve it.

Comment (nickvergessen) [inline apps/files_sharing/api/server2server.php]: It's null, when we don't mount the remote share now. This is when someone inserts a cloudId in the share sidebar.
In this case there needs to be a "accept please" message, which is exactly the case which we want.

But all good with my fix I think

**Comment by nickvergessen** [comment]
Wrong time for this joke oracle:
```
08:38:08 ownCloud is not installed - only a limited number of commands are available
08:38:09 Oracle connection could not be established
08:38:09  -> ORA-01017: invalid username/password; logon denied Check environment: ORACLE_HOME=/usr/lib/oracle/xe/app/oracle/product/10.2.0/server ORACLE_SID=XE LD_LIBRARY_PATH=/usr/lib/oracle/xe/app/oracle/product/10.2.0/server/lib: NLS_LANG=AMERICAN_AMERICA.AL32UTF8 tnsnames.ora is readable
```

repushing

**Comment by nickvergessen** [comment]
postgres failure is the known hick up where the "md5" is unequal to the string object "md5"

**Comment by PVince81** [comment]
:+1:

**Pull request merged**